SleepIQ component with sensor and binary sensor platforms

technicalpickles opened the pull request:
**Description:**

My bed has an API, so obviously I want to integrate it into homeassistant :D SleepNumber beds have an addon called [SleepIQ](http://www.sleepnumber.com/sn/en/sleepiq-sleep-tracker). It tracks sleep over time, and in particular, it does know if you are in bed or not, and it knows your 'SleepNumber' (firmness of the mattress). There are controls for changing the incline of the frame, and firmness, but it requires using bluetooth near the bed, and doesn't seem to be exposed by the API (I wrote the library sleepyq for interacting it, and made some notes: https://github.com/technicalpickles/sleepyq).

This is my first component, and my first real python, so definitely looking for feedback both in terms of style & implementation details. I started discussion about this over in https://community.home-assistant.io/t/feedback-on-component-for-sleepiq-beds/3132

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant/pull/2949

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
sleepiq:
  login: !secret sleepiq_login
  password: !secret sleepiq_password
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices, web services, or a:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

technicalpickles committed:
First pass at SleepIQ component with sensor and binary sensor platforms

Teagan42 commented on homeassistant/components/sensor/sleepiq.py:
```diff
@@ -0,0 +1,86 @@
+
+    def __init__(self, sleepiq_data, bed_name, side, sensor_type):
+        """Initialize the sensor."""
+        self.client_name = 'SleepNumber'
```
Magic String - use constant

Teagan42 commented on homeassistant/components/sensor/sleepiq.py:
```diff
@@ -0,0 +1,86 @@
+
+        self._sleeper = status['sleeper']
+
+        if self.type == 'sleep_number':
```
More magic strings

Teagan42 commented on homeassistant/components/binary_sensor/sleepiq.py:
```diff
@@ -0,0 +1,15 @@
+"""
```
Not sure the point of this file - know it's WIP but need more info to even understand

technicalpickles replied:
I've based this on existing components. I'm not a fan of it myself, but most of them just say they are support for this and that. Examples:

* https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/ecobee.py
  * https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/ecobee.py
* https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/fitbit.py

Anything in particular I can answer?

Teagan42 replied:
You're not defining any sensors in the file... that's why I'm not sure

technicalpickles replied:
@Teagan42 yet 😅

This started as a single sensor platform, and I hadn't factored it out to have a binary_sensor for is_in_bed yet. I pushed that up earlier.

Teagan42 commented on homeassistant/components/sleepiq.py:
```diff
@@ -0,0 +1,88 @@
+import logging
+from datetime import timedelta
+from requests.exceptions import ConnectionError as ConnectError, \
```
`from requests.exceptions import (ConnectionError, HttpError, Timeout)`

Teagan42 commented on homeassistant/components/sleepiq.py:
```diff
@@ -0,0 +1,88 @@
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=30)
+
+SENSOR_TYPES = {
+    'sleep_number': 'SleepNumber',
```
Lots of magic strings - make constants for these

Teagan42 commented:
Utilize constants instead of re-typing all those strings. Make sure your add unit tests or add it to the `.coveragerc` (unit tests preferred)

technicalpickles commented:
Thanks for all the feedback 😀 

I was having problems getting the component to load at all, so I pushed this a little earlier than I would have normally so I could ask. Then I fixed the problems 😳 (basically trying to use `python setup.py develop` on hass, and then running `hass` was loading the wrong version for some reason).

As soon as I did that, I found the sleepyq wrapper I wrote stopped working. I think they changed something on their backend, but I'm still tracking it down.

technicalpickles committed:
Update sleepyq dependency

technicalpickles committed:
config is top level config, not scoped to the domain

technicalpickles committed:
address some balloobbot feedback

technicalpickles committed:
Make sure discovery is imported, and return true from setup

technicalpickles committed:
fix variable renamed missed. comment out binary sensors until they move

technicalpickles committed:
implement binary sensor

technicalpickles committed:
sufficient whitespace

technicalpickles commented on homeassistant/components/sensor/sleepiq.py:
```diff
@@ -0,0 +1,86 @@
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
```
Fixed this by adding to `ICON` to the top.

technicalpickles commented on homeassistant/components/sensor/sleepiq.py:
```diff
@@ -0,0 +1,86 @@
+    if discovery_info is None:
+        return
+
+    data = sleepiq.DATA
```
This was fixed.

technicalpickles commented on homeassistant/components/sleepiq.py:
```diff
@@ -0,0 +1,88 @@
+import logging
```
This was fixed.

technicalpickles committed:
reduce conditionals in sensor by exposing side status

technicalpickles committed:
unused code

technicalpickles committed:
try to reduce code for sides of the bed

technicalpickles commented:
> As soon as I did that, I found the sleepyq wrapper I wrote stopped working. I think they changed something on their backend, but I'm still tracking it down.

This turned out to me not expecting the right things in `config` in `setup` (it's the entire config, not scoped to the component).

Aside from that, the functionality seems to be working. There's definitely some cleanup to do:

* [ ] magic strings everywhere
* [ ] there's a lot of replication between the binary sensor and sensor. not sure how to factor that out yet
* [ ] the "Is In Bed" sensor seems to always show the icon when ungrouped, rather than the value
* [ ] tests would be good, but not sure I'll be able to. can anyone point me to similar component tests where the component is making http requests?
* [ ] address things from @balloobbot 

And other things that will make this more usable in general:

* [ ] validate configuration
* [ ] better error handling (in sleepyq)

Some other improvements suggested from over on https://community.home-assistant.io/t/feedback-on-component-for-sleepiq-beds/3132/2 cc @robbiet480 

* [ ] include state attributes. I'm not sure I know how these'd be used yet though.
* [ ] possibly introduce a new sensor class like occupancy? or maybe a more generic one like yes or no (since the binary sensor is called "is in bed")
* [ ] customize the binary sensor icon

technicalpickles committed:
Update to sleepyq that raises on non-200 status codes

technicalpickles committed:
Integrate sleepyq==0.5 which has improvements to help make this cleaner

It supports multiple beds now (theroetically? someone want to buy me a
bed to test?). It also returns domain specific objects, so there's less
magic strings for key acceses. Between those too things, was able to
clean up quite a bit.

technicalpickles committed:
use standard config key (username instead of login), use constants, and validate config

technicalpickles committed:
Add error handling

technicalpickles committed:
Remove client-Name from sensors, since not really used except for formatting name

technicalpickles committed:
Remove custom icon for binary_sensor, so it changes based on state

technicalpickles committed:
Add sleepiq to coveragerc

technicalpickles committed:
Run script/gen_requirements_all.py

technicalpickles committed:
Fix import of sleepiq to be only during setup. Change SleepIQData to take a client rather than the username/password

technicalpickles committed:
address some lint errors

technicalpickles committed:
Merge branch 'dev' into sleepiq

technicalpickles commented on homeassistant/components/binary_sensor/sleepiq.py:
```diff
@@ -0,0 +1,15 @@
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.sleepiq/
+"""
+from homeassistant.components import sleepiq
```
This has been addressed.

technicalpickles commented on homeassistant/components/binary_sensor/sleepiq.py:
```diff
@@ -0,0 +1,15 @@
+    """Setup the SleepIQ sensors."""
+    if discovery_info is None:
+        return
+
```
This has been addressed (I think?)

technicalpickles committed:
address more balloobbot feedback

technicalpickles committed:
fix long line. copy _LOGGER convention from bloomsky

technicalpickles committed:
address lint errors

technicalpickles committed:
fix unused variables

technicalpickles committed:
fix pydocstyle failures

technicalpickles committed:
Merge branch 'dev' into sleepiq

technicalpickles committed:
add base sensor class, then use multiple inheritence for binar sensor

technicalpickles committed:
Address pylint and pydocstyle errors

technicalpickles committed:
fix flake8 errors

technicalpickles committed:
really fix it?

technicalpickles commented:
I'm feeling pretty good about this PR, so it's ready for 👀 . I also removed the WIP title.

I addressed most of the items in the PR template. I also did the cleanup and usability things I listed out. I'm going to hold off on further improvements and if it's okay, tests for now.

The only thing I have left is adding documentation to homeassistant.io.

technicalpickles committed:
wrong constant

technicalpickles committed:
remove unused code, since subclasses handle this now

robbiet480 commented:
@technicalpickles I've gotta do some UI work tonight for the fan component, i'll try to throw in the new occupancy SENSOR_CLASS. You can go ahead and set the class in your attributes now.

robbiet480 commented:
@technicalpickles I'm sorry but this can't be merged without tests. New components always require tests, no matter how simple/complex they are. We have t-minus 10 hours until we cut 0.27. It would be great to see this make it in, but please don't stress if it doesn't. It will go out in 0.28 if not. :smile:

technicalpickles commented:
@robbiet480 looking at the [PR template](https://github.com/home-assistant/home-assistant/blob/dev/.github/PULL_REQUEST_TEMPLATE.md) it sounds like code that interacts with devices or web services don't need tests. If that's not the case, it's be good to make that more clear.

Any pointers for testing this would be appreciated, i.e. tests for components that are similar in they poll a web service.

Teagan42 commented:
@technicalpickles You're correct - if it talks to a device, technically, unit tests are not required for talking to the device.

However, initial platform setup and configuration validation don't need to talk to device.

It's also possible, if you want some extra 🍪 , to mock the response from the library/device and write unit tests anyway.

technicalpickles committed:
Start some tests

technicalpickles committed:
Fix whitespace

technicalpickles committed:
start adding tests, with fixtures

technicalpickles committed:
Don't exclude from coverage

technicalpickles committed:
Fix invalid login setup test

technicalpickles committed:
fix accidentally encoding the fixture as a string rather than the json types they should be

technicalpickles committed:
Add actual tests for sensor & binary_sensor

technicalpickles committed:
Fix flake8 errors, refactor fixture loading to one place

technicalpickles committed:
Fix long long

technicalpickles committed:
Merge branch 'dev' into sleepiq

technicalpickles commented:
I've added tests, but they probably aren't great. I also wrote documentation over in https://github.com/home-assistant/home-assistant.io

I'm sorting out the last few lint errors, but should be ready otherwise.

technicalpickles commented:
Is there any other feedback on this? I'm not sure what the release cycle is, but I just want to make sure I have time to address it before the next release.

robbiet480 commented:
@technicalpickles 0.28 won't be released for another 2 weeks from last Saturday, so you've got plenty of time. We'll take a look at this PR again well before then.

robbiet480 commented:
@technicalpickles This looks good to merge for me. However, I owe you a new `sensor_class`. Want to pick out an occupied and not occupied icon from [MDI](https://materialdesignicons.com) and i'll get it integrated into the frontend?

technicalpickles commented:
@robbiet480 great, thanks!

I actually was looking for icons before, but didn't see any that really seemed great. I was hoping for like a room icon (ie looking at a corner) with and without a person.

Maybe account-outline for unoccupied and account for occupied? That would make it pretty similar to the normal dot and filled dot, but be fore a person. Or home-outline and home?

robbiet480 commented:
Gonna go with home/home-outline. Will be done in a few.

robbiet480 commented:
https://github.com/home-assistant/home-assistant/pull/3176

Just add something to your code like:

```python
@property
    def sensor_class(self):
        """Return the class of this sensor."""
        return "occupancy"
```

technicalpickles committed:
use occupancy sensor class

technicalpickles committed:
Merge branch 'dev' into sleepiq

technicalpickles commented:
@robbiet480 done, thanks!

Teagan42 commented:
@robbiet480 Looks good to me - you?

pvizeli commented on homeassistant/components/binary_sensor/sleepiq.py:
```diff
@@ -0,0 +1,71 @@
+        return self._state is True
+
+    @property
+    def state(self):
```
Do not overwrite this function. See on base class `BinarySensorDevice` that use only the `is_on` function to get the state.

technicalpickles replied:
I'm overwritting it to make sure it uses BinarySensorDevice's state, rather than SleepIQSensor.

pvizeli replied:
Push that to `is_on`. Show on BinarySensorDevice to show how state work.

technicalpickles replied:
> Show on BinarySensorDevice to show how state work.

I don't follow what you mean here.

I already have a `is_on` in this class. My concern is that if something does check the state on this, it would get `SleepIQSensor#state` instead of `BinarySensorDevice#state`. That would return `"True"` and `"False"` rather than `"on"` and `"off"`.

technicalpickles replied:
Without an overriden `state`, this happens:

![](https://cl.ly/20151Y12450U/Image%202016-09-11%20at%209.46.14%20AM.png)

The state should be `off`, not `False`.

pvizeli commented:
fix my comment and I'm happy to :dolphin:

technicalpickles commented:
Looks like I missed 0.28 😕

pvizeli commented:
I see. Show on Homematic how that should works. You don't need a state in base 'SleepIQSensor' so you owerwrite core things on other components. All other comoonent have here own state. Only the sensor use a own state method. So if you remove state on base. Can you delete that on binary and you add state to sensor component class for you sleep.

technicalpickles committed:
Move state and icon down into sleepiqsensor base class

technicalpickles commented:
I'm having a hard time understanding the reason for them, but I've applied what I understand the feedback to be and confirmed the functionality is intact.

pvizeli commented:
Perfect. On Sensor component do you have forget to delete the icon static variable or add you icon function (thats is allowed). Please use also volutuous for config validation (see on other component or homematic) And then is ready to merge. Good work :+1:

pvizeli commented:
Like that https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/homematic.py#L123-L136

technicalpickles committed:
Merge branch 'dev' into sleepiq

technicalpickles committed:
Add back icon for sleepnumbersensor

technicalpickles committed:
use voluptuous

pvizeli commented:
Yeah, looks good :+1: pass all test and ist done. I'll change the order of class/setup (first setup function and later the class) in your code after merge, so you don't need to do that if you don't want to do self.

technicalpickles commented:
@pvizeli I made the changes, but now I have a test failure with voluptuous instead of using `validate_config` because I had test cases for the missing configuration. Is there a voluptuous way of testing config, or should I just remove those?

pvizeli commented:
use `bootstrap._setup_component` See i.e. notify file test.

technicalpickles committed:
use _setup_component to confirm validation

technicalpickles commented:
@pvizeli added that, thanks!

pvizeli commented on tests/components/test_sleepiq.py:
```diff
@@ -57,13 +58,13 @@ def test_setup_login_failed(self, mock):
         del self.config['sleepiq']['username']
-        sleepiq.setup(self.hass, self.config)
-        self.assertFalse(sleepiq.setup(self.hass, self.config))
+        assert not bootstrap._setup_component(self.hass, sleepiq.DOMAIN, self.config)
```
You need copy config and del from copy or you del from main config and in next unit test it is also deleted

pvizeli commented:
I've made a clean PR without merges and make some code cleanups. Can you test it? #3390

technicalpickles commented:
@pvizeli if you wanted a clean branch, I could have made one. You squashed the commits, so I can't easily see what changes you made or why so I can learn, and it also don't count as a contribution from me.

pvizeli commented:
You are the mentainer and you stay in release note. On Home-Assistant developer guide is written that PR should be rebased and squash, thats all. For new PR we have upstream rights to help but that was before github have this feature. I want only help you and spend some hours for that.

balloob commented:
Sorry @technicalpickles, I guess we're not all GitHub guru's like you. 

@pvizeli some tips for next time:

You can commit directly to his branch on his fork with the new [improved collaboration features](https://github.com/blog/2247-improving-collaboration-with-forks).

If you use the 'squash and merge' button on GitHub, the original author will remain intact. [See example](https://github.com/home-assistant/home-assistant/commit/68def21615acccdfbaaff6b0855fcf980fc07983)

pvizeli commented:
I'm sorry. I've only want to help

balloob commented:
Don't worry @pvizeli, we're all learning here. It happens

technicalpickles commented:
I'm sorry if if my last comment came of short or antagonistic. I didn't mean it that way, it's just that this PR has been quite the voyage for me.

I'm new to Home Assistant development, and actually Python too. It wasn't easy trying to pick it all up at once, but I learned a ton along the way. I was mostly disappointed in not being able to get it merged myself. I wanted to know what changes would be needed to be merged for my own learning and also to contribute better components in the future, that is all. I will gladly trade some Git and GitHub expertise for Python and Home Assistant expertise 😀 

Home Assistant is probably the best managed open source project I've watched and participated in in some time. Thanks for everyone that left feedback and thanks to @pvizeli in particular for cleaning up the PR and getting it merged for 0.29.
